cmd/server/server.go

```diff
@@ -0,0 +1,211 @@
+	"github.com/gardener/etcd-backup-restore/pkg/snapstore"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+const (
+	envStorageContainer             = "STORAGE_CONTAINER"
+	defaultLocalStore               = "default.etcd.bkp"
+	backupFormatVersion             = "v1"
+	defaultServerPort               = 8080
+	defaultName                     = "default"
+	defaultInitialAdvertisePeerURLs = "http://localhost:2380"
+)
+
+var logger = logrus.New()
+
+// configuration is struct to hold configuration for utility
+type configuration struct {
+	schedule              string
+	etcdEndpoints         string
+	storageProvider       string
+	maxBackups            int
+	etcdConnectionTimeout int
+}
+
+var (
+	port int
+
+	//restore flags
+	restoreCluster      string
+	restoreClusterToken string
+	restoreDataDir      string
+	restorePeerURLs     []string
+	restoreName         string
+	skipHashCheck       bool
+	storageProvider     string
+)
+
+// NewServerCommand create cobra command for snapshot
+func NewServerCommand(stopCh <-chan struct{}) *cobra.Command {
+	config := &configuration{}
+	var command = &cobra.Command{
+		Use:   "server",
+		Short: "start the http server with backup scheduler.",
+		Long:  `Server will keep listening for http request to deliver its functionality through http endpoins.`,
+		Run: func(cmd *cobra.Command, args []string) {
+
+			clusterUrlsMap, err := types.NewURLsMap(restoreCluster)
+			if err != nil {
+				logger.Fatalf("failed creating url map for restore cluster: %v", err)
+			}
+			peerUrls, err := types.NewURLs(restorePeerURLs)
+			if err != nil {
+				logger.Fatalf("failed parsing peers urls for restore cluster: %v", err)
+			}
+
+			options := &restorer.RestoreOptions{
+				RestoreDataDir: restoreDataDir,
+				Name:           restoreName,
+				ClusterURLs:    clusterUrlsMap,
+				PeerURLs:       peerUrls,
+				ClusterToken:   restoreClusterToken,
+				SkipHashCheck:  skipHashCheck,
+			}
+
+			etcdInitializer := initializer.NewInitializer(options, storageProvider, logger)
+			handler := &server.HTTPHandler{
+				Port:            port,
+				EtcdInitializer: *etcdInitializer,
+				Logger:          logger,
+			}
+			logger.Info("Regsitering the http request handlers...")
+			handler.RegisterHandler()
+			logger.Info("Starting the http server...")
+			go handler.Start()
+
+			defer handler.Stop()
+
+			ss, err := getSnapstore(storageProvider)
+			if err != nil {
+				logger.Fatalf("Failed to create snapstore from configured storage provider: %v", err)
+			}
+			logger.Info("Created snapstore from provider: %s", storageProvider)
+			ssr, err := snapshotter.NewSnapshotter(
+				config.etcdEndpoints,
+				config.schedule,
+				ss,
+				logger,
+				config.maxBackups,
+				time.Duration(config.etcdConnectionTimeout))
+			if err != nil {
+				logger.Fatalf("Failed to create snapshotter: %v", err)
+			}
+			for {
+				logger.Infof("Probing etcd...")
+				select {
```

Comment:
We need to avoid busy loops.

https://www.reddit.com/r/golang/comments/6ec9o2/effect_of_infinite_loop_on_cpumem/

Reply:
As per my understanding, this is not a busy loop. This is polling mechanism. Since we don't get any event from etcd running as separate process, we have to poll for it to check whether it's up or not and also for safe termination we need `stopCh`.  If you are wondering about `default` case behaviour, then please understand the `probeEtcd` call one level deep. Internally, it's a socket or to be precise IO blocking call with timeout. CPU will hit the socket, and wait for interrupt from n/w device. So, indirectly speaking, CPU is ideal in both stopCh case and default case. You can also verify yourself using `top`  by running simple `etcdbrctl server` command (Don't start `etcd` otherwise it will exit out of for).